src/lib.rs

```diff
@@ -102,50 +104,20 @@ mod parser;
 const VERSION: &str = "0.0.1";
 const LANG: &str = "rust";
 
-#[doc(hidden)]
-pub trait ConnectionState: private::Sealed {}
-mod private {
-    pub trait Sealed {}
-    impl Sealed for super::NotConnected {}
-    impl Sealed for super::Authenticated {}
-    impl Sealed for super::Connected {}
-    impl Sealed for dyn super::OptionState {}
-}
-
-impl ConnectionState for NotConnected {}
-impl ConnectionState for Authenticated {}
-impl ConnectionState for Connected {}
-
-/// A NATS connection.
-#[derive(Debug)]
-pub struct Connection<S: ConnectionState> {
-    options: Options,
-    state: S,
-}
-
-#[doc(hidden)]
-pub trait OptionState {}
-impl OptionState for NotConnected {}
-impl OptionState for Authenticated {}
-
-// General Options
-impl<S> Connection<S>
-where
-    S: OptionState + ConnectionState,
-{
+impl Options {
```

Comment:
You are doing a blanket one, where I also had the ability to distinguish an authorized state too which is a type of option but allows the compiler to not allow to use multiple ones.

Reply:
is the change you're requesting that authentication is only settable once, in a way that is enforced at compile time?